src/language-js/needs-parens.js

```diff
@@ -270,10 +283,34 @@ function needsParens(path, options) {
         case "AwaitExpression":
         case "TSNonNullExpression":
         case "UpdateExpression":
+          // [prettierx] parenSpace option support (...)
+          // Logical and Binary expressions already got their parens if parent is UnaryExpression
+          // if (
+          //   (parent.type === "UnaryExpression" &&
+          //   (node.type === "LogicalExpression" ||
+          //     node.type === "BinaryExpression")) &&
+          //     !(node.extra && node.extra.parenthesized)
+          // ) {
+          //   return false;
+          // }
```

Comment:
🎉 Yeah !!

I *may* update these comments a little and do some more testing.

Reply:
> I _may_ update these comments a little and do some more testing.

Go for it! Before I commented the additional logic out, it was causing all sorts of problems all over the place, getting rid of parentheses in cases where they shouldn't have been; one specific example I remember being

```js
!(
  condition1 &&
  condition2
)
```

becoming

```js
!
  condition1 &&
  condition2

```